src/aws-codeartifact/domain.ts

```diff
@@ -0,0 +1,335 @@
+
+  /**
+   * The KMS key used to encrypt the Domain
+   */
+  readonly encryptionKey?: IKey;
+
+  /**
+   * Adds a statement to the Codeartifact domain resource policy.
+   * @param statement The policy statement to add
+   */
+  addToResourcePolicy(statement: PolicyStatement): AddToResourcePolicyResult;
+
+  /**
+   * Grants permissions to the specified grantee on this CodeArtifact domain.
+   *
+   * It handles both same-environment and cross-environment scenarios:
+   * - For same-environment grants, it adds the permissions to the principal or resource.
+   * - For cross-environment grants, it adds the permissions to both the principal and the resource.
+   *
+   * @param grantee - The principal to grant permissions to.
+   * @param actions - The actions to grant. These should be valid CodeArtifact actions.
+   */
+  grant(grantee: IGrantable, ...actions: string[]): Grant;
+
+  /**
+   * Grants contribute permissions to the specified grantee on this CodeArtifact domain.
+   *
+   * @param grantee - The principal to grant contribute permissions to.
+   */
+  grantContribute(grantee: IGrantable): Grant;
+}
+
+/**
+ * A new or imported Codeartifact Domain.
+ */
+abstract class DomainBase extends Resource implements IDomain {
+  /**
+   * The ARN (Amazon Resource Name) of the CodeArtifact domain.
+   */
+  public abstract readonly domainArn: string;
+
+  /**
+   * The name of the CodeArtifact domain.
+   */
+  public abstract readonly domainName: string;
+
+  /**
+   * The AWS KMS encryption key associated with the domain, if any.
+   */
+  public abstract readonly encryptionKey?: IKey;
+
+  /**
+   * The ARN of the key used to encrypt the Domain
+   */
+  public abstract readonly domainEncryptionKey?: string;
+
+  /**
+   * The AWS account ID that owns the domain.
+   */
+  public abstract readonly domainOwner: string;
+
+  /**
+   * Optional policy document that represents the resource policy of this domain.
+   */
+  protected abstract readonly policy?: PolicyDocument;
+
+  /**
+   * Adds a statement to the Codeartifact domain resource policy.
+   * @param statement The policy statement to add
+   */
+  public addToResourcePolicy(statement: PolicyStatement): AddToResourcePolicyResult {
+    const stack = Stack.of(this);
+
+    if (!this.policy) {
+      Annotations.of(stack).addWarningV2(
+        'NoResourcePolicyStatementAdded',
+        `No statements added to imported resource ${this.domainArn}.`,
+      );
+      return { statementAdded: false };
+    }
+
+    this.policy.addStatements(statement);
+    return { statementAdded: true, policyDependable: this.policy };
+  }
+
+  private isCrossEnvironmentGrantee(grantee: IGrantable): boolean {
+    if (!principalIsOwnedResource(grantee.grantPrincipal)) {
+      return false;
+    }
+    const thisStack = Stack.of(this);
+    const identityStack = Stack.of(grantee.grantPrincipal);
+    return thisStack.region !== identityStack.region || thisStack.account !== identityStack.account;
```

Comment:
Both `region` and `account` need to consider the possibility of tokens.

It is good to use `Token.compareStrings`.

Reply:
Didn't know about that method. Nice :)